### ISSUE TITLE
[CURATOR-504] Race conditions in LeaderLatch after reconnecting to ensemble

### ISSUE THREAD

**Issue by jira-importer** [apache/curator#1024]
<p>We use LeaderLatch in a lot of places in our system and when ZooKeeper ensemble is unstable and clients are reconnecting to logs are full of messages like the following:</p>

<p><tt><span class="error">&#91;2017-08-31 19:18:34,562&#93;</span><span class="error">&#91;ERROR&#93;</span><span class="error">&#91;org.apache.curator.framework.recipes.leader.LeaderLatch&#93;</span> Can't find our node. Resetting. Index: -1 {</tt>}</p>

<p>According to the <a href="https://github.com/apache/curator/blob/4251fe328908e5fca37af034fabc190aa452c73f/curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderLatch.java#L529-L536" class="external-link" target="_blank" rel="nofollow noopener">implementation</a>, this can happen in two cases:</p>
<ul>
	<li>When internal state `ourPath` is null</li>
	<li>When the list of latches does not have the expected one.</li>
</ul>

<p>I believe we hit the first condition because of races that occur after client reconnects to ZooKeeper.</p>
<ul>
	<li>Client reconnects to ZooKeeper and LeaderLatch gets the event and calls reset method which set the internal state (`ourPath`) to null, removes old latch and creates a new one. This happens in thread "Curator-ConnectionStateManager-0".</li>
	<li>Almost simultaneously, LeaderLatch gets another even NodeDeleted (<a href="https://github.com/apache/curator/blob/4251fe328908e5fca37af034fabc190aa452c73f/curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderLatch.java#L543-L554" class="external-link" target="_blank" rel="nofollow noopener">here</a>) and tries to re-read the list of latches and check leadership. This happens in the thread "main-EventThread".</li>
</ul>

<p>Therefore, sometimes there is a situation when method `checkLeadership` is called when `ourPath` is null.</p>

<p>Below is an approximate diagram of what happens:</p>

<p><span class="image-wrap" style=""><img src="https://issues.apache.org/jira/secure/attachment/12957174/12957174_51868597-65791000-231c-11e9-9bfa-1def62bc3ea1.png" height="150" width="1261" style="border: 0px solid black" /></span></p>

---
<details><summary><i>Originally reported by <a title="Yuri Tceretian" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian">yuri.tceretian</a>, imported from: <a href="https://issues.apache.org/jira/browse/CURATOR-504" target="_blank">Race conditions in LeaderLatch after reconnecting to ensemble</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Zili Chen" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tison">tison</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2023-08-16T10:43:05+01:00
<li><b>imported</b>: 2025-01-21
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>Hey <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian" class="user-hover" rel="yuri.tceretian">Yuri Tceretian</a> - nice to see you over "here" <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/biggrin.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>This sounds like it might be related to something we're about to fix: <a href="https://issues.apache.org/jira/browse/CURATOR-498" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/CURATOR-498</a> - please look that over and see if it corresponds.</p>

<p>Update: Looking at the graphic posted it looks a lot like what we're seeing in <a href="https://issues.apache.org/jira/browse/CURATOR-498" title="Protected Mode creation can mistake closing session&#39;s node causing problems for many recipes such as LeaderLatch" class="issue-link" data-issue-key="CURATOR-498"><del>CURATOR-498</del></a>.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian">yuri.tceretian</a>:</i>
<p>Hi <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt" class="user-hover" rel="randgalt">Jordan Zimmerman</a>, yes I saw that issue and from its description, I decided that it was a slightly different one. In this case, it's simple races within LeaderLatch, reconnect happens right after the connection is lost, ie within one session. Honestly, I did not read all the comments. So, if you think that this is a duplicate, feel free to close it.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian">yuri.tceretian</a>:</i>
<p>Also, there is at least one potentially dangerous consequence: (see diagram above) If latch X becomes the leader (because zNode was not deleted due to races) but the instance has latch Y, then there can be a situation when "ghost"  latch is considered a leader and therefore there is no leader, until the session is closed and the "ghost" latch disappears. </p>

<p> </p>

<p>Another founding: If something from outside deletes the leader's zNode, the current leader does not get notified but candidates do, and this lead to a situation when there are many leaders. Not sure this is a bug because, but I think it is worth mentioning.  </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<blockquote><p>I believe we hit the first condition because of races that occur after client reconnects to ZooKeeper.<br/>
Client reconnects to ZooKeeper and LeaderLatch gets the event and calls reset method which set the internal state (`ourPath`) to null, removes old latch and creates a new one. This happens in thread "Curator-ConnectionStateManager-0".<br/>
Almost simultaneously, LeaderLatch gets another even NodeDeleted (here) and tries to re-read the list of latches and check leadership. This happens in the thread "main-EventThread".</p></blockquote>

<p>This isn't really a race. There's not much that can be done about it. The LeaderLatch doesn't know what its node is until the background creation occurs. The only reason that a NodeDeleted event would be received is if the previous sequence node gets deleted. So, that's a rare case but could happen. The end result, though, is merely that the Latch resets. This is noisy but not race-y as the end result is fine. A test case should be written that simulates this just in case, though. There might be an optimization to detect that the latch is in this state - i.e. it's waiting for the node creation result - but I think that would be difficult and error prone.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<blockquote><p>Also, there is at least one potentially dangerous consequence: (see diagram above) If latch X becomes the leader (because zNode was not deleted due to races) but the instance has latch Y, then there can be a situation when "ghost"  latch is considered a leader and therefore there is no leader, until the session is closed and the "ghost" latch disappears. </p></blockquote>

<p>I don't see how the latch's zNode would not be deleted. Every node that the latch creates goes through `setNode(path)` which deletes any previous zNode.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<blockquote><p>Another founding: If something from outside deletes the leader's zNode, the current leader does not get notified but candidates do, and this lead to a situation when there are many leaders. Not sure this is a bug because, but I think it is worth mentioning.  </p></blockquote>

<p>This is defined invariant in Curator. We have Tech Note on it: <a href="https://cwiki.apache.org/confluence/display/CURATOR/TN7" class="external-link" target="_blank" rel="nofollow noopener">https://cwiki.apache.org/confluence/display/CURATOR/TN7</a> - it would be very expensive and cumbersome to add watchers to lock nodes just to guard against this.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p> <span class="image-wrap" style=""><img src="https://issues.apache.org/jira/secure/attachment/12957195/12957195_Screen+Shot+2019-01-31+at+10.26.59+PM.png" style="border: 0px solid black" /></span> </p>

<p>Note: when <tt>ourPath</tt> becomes Y, ZNode X gets deleted. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>I wrote this test (as part of <a href="https://issues.apache.org/jira/browse/CURATOR-498" title="Protected Mode creation can mistake closing session&#39;s node causing problems for many recipes such as LeaderLatch" class="issue-link" data-issue-key="CURATOR-498"><del>CURATOR-498</del></a>) that I believe simulates the scenario describe in this issue. The test passes showing that there is no race. Thopughts <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian" class="user-hover" rel="yuri.tceretian">Yuri Tceretian</a>?</p>

<p><a href="https://github.com/apache/curator/blob/CURATOR-498/curator-recipes/src/test/java/org/apache/curator/framework/recipes/leader/TestLeaderLatch.java#L73" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/blob/CURATOR-498/curator-recipes/src/test/java/org/apache/curator/framework/recipes/leader/TestLeaderLatch.java#L73</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian">yuri.tceretian</a>:</i>
<p>I agree that "ghost" latch is not possible due to "races" I mentioned in the issue. Apparently I put the "delete znode" command, in wrong place. As you pointed out above, It is sent a little bit later and deletes node X, when ourPath is set to Y.</p>

<p><span class="image-wrap" style=""><img src="https://issues.apache.org/jira/secure/attachment/12957686/12957686_XP91JuD048Nl_8h9NZpH01QZJMfCLewjfd2eQNfOsR6GuApPNV.png" height="132" width="1199" style="border: 0px solid black" /></span>  </p>

<p>The test above does not seem to simulate the original issue, which happens inside the instance of leader latch.</p>

<p>Yes, it is not a big deal, because it is eventually consistent. Moreover, usually, it does not happen. We observe this issue when a host where the process is running, is under load.</p>

<p>Although it is just a matter of a few extra requests to be sent by a leader latch instance while stabilizing its internal state, it becomes quite noisy in a situation when there are a lot of latches per a connection, and together with a couple of tree cache instances can simply cause a big hit on request queue. </p>

<p>I think that this is a minor bug, and it just makes thing noisier when the connection is not stable. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>I'm not sure how to make it less noisy without compromising consistency. ZK does not give us a lot of options API-wise.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian" class="user-hover" rel="yuri.tceretian">Yuri Tceretian</a>: I've had an idea for a while of a circuit breaker style <tt>ConnectionStateListener</tt>. It would proxy any ConnectionStateListeners used by Curator recipe/classes such that when the connection is lost the circuit would open for a period of time and, while open, ignore any changes in state. After the time period expires the circuit would close and send whatever the current connection state is. This way, if the connection is going up/down/up/down/up/down, the application would only see the first down and then N ms later hopefully the connection is repaired and the application would only see the reconnection.</p>

<p>Thoughts?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian">yuri.tceretian</a>:</i>
<p>I like the idea very much! It can be configurable, and probably with some random jitter (also configurable). Therefore, If a host had intermittent connection issues, applications running on that host do not hit the ZK simultaneously.</p>

<p> </p>

<p>Regarding the issue, I wonder if it is possible to do the following without hurting consistency:</p>
<ul>
	<li>add a boolean field that is set in `reset` method and reset in the callback that is called after a node is created.</li>
	<li>in the method checkLeadership, when the current state is null, call reset if the flag is reset.</li>
</ul>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p><a href="https://issues.apache.org/jira/browse/CURATOR-505" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/CURATOR-505</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian">yuri.tceretian</a>:</i>
<p>I think this is more serious than I initially believed. I encountered a situation when an instance of LeaderLatch got into a loop of creating\deleting znodes and it stopped only after we restarted the service.  </p>

<p>The <a href="https://gist.github.com/yuri-tceretian/1f5ea4f37f574cae7ec01d75f5c11f9a" class="external-link" target="_blank" rel="nofollow noopener">gist</a> (can't attach files to the issue) contains a piece of logs from the client that shows a huge amount of requests produced by one client. In one of lines you can find a list of znodes that the latch's znode contained at that time:</p>

<p><tt>'_c_7c38ac16-5844-41b8-a8af-cd4f4624c3e9-latch-0002560578</tt><br/>
<tt>'_c_500ff830-754d-42a8-b695-6a1d6422a306-latch-0000000136</tt><br/>
<tt>'_c_33a67632-037b-4b30-9623-dfdf60ddbbe9-latch-0002560577</tt><br/>
<tt>'_c_6d2fdd32-8e16-453a-a953-dcc06fccc380-latch-0000000130</tt><br/>
<tt>'_c_f37f97c2-8f2d-4bf6-b363-dfcd5e480fa3-latch-0002560576</tt><br/>
<tt>'_c_c83883a8-2d0a-46d5-afec-ff67f9f6d5d0-latch-0000000133</tt><br/>
<tt>'_c_1bc15cc0-c46e-4270-9502-9e0673eec763-latch-0000000128</tt></p>

<p>as you can see there is a node with index 0002560576, which means that there were created 2.6 million of znodes (for about 2 hours) (3 znodes with a big index can be explained as that there were 2 leader latches that shared same connection and pointed to the same znode). I believe the problem started happening after the client timed out and reconnected to ZooKeeper.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>That gist has a few Curator lines and then a ton of ZooKeeper lines - is that filtered? It would be nice to have more of the Curator log lines. Also, this still may be the issue fixed in 4.2.0. Is this using 4.2.0 or an earlier version?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian">yuri.tceretian</a>:</i>
<p>That was 4.1.0 (still). We are in the process of migrating to 4.2.0. If you mean circuit breaker in 4.2.0, I do not think it would help because that was not a flapping connection. </p>

<p>You're correct about logs, there were only ZooKeeper's logs because Curator's debug logs were not enabled. I will try to reproduce it and find more logs.  </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>No, not circuit breaker. 4.2.0 fixes a serious bug that might be the source of the problem (I'm not sure thought).</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tison">tison</a>:</i>
<p>I prepare a patch to release this situation in <a href="https://github.com/apache/curator/pull/430" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/430</a> .</p>

<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=yuri.tceretian" class="user-hover" rel="yuri.tceretian">Yuri Tceretian</a> I'd like to know whether this is a livelock or one shot race which after the last getChildren<span class="error">&#91;1&#93;</span><span class="error">&#91;2&#93;</span> returned, it will converge.</p>

<p>I try to simulate the time order and it seems this case should not be a livelock, but if it's not one, how can you get millions of competitors.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tison">tison</a>:</i>
<p>I think I mix the original logic and patch I made. If we reset on preceding node not found, it's a live lock.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tison">tison</a>:</i>
<p>Also I add several debug log points so that if the issue remains, we can be more clear how it happens.</p>

**Comment by jira-importer** [comment]
<i>[Originally related to: <a href="https://github.com/apache/curator/issues?q=in%3Atitle%20CURATOR-505">CURATOR-505</a>]</i>
